### ISSUE TITLE
Include postal codes for localities

### ISSUE THREAD

**Issue by NickStallman** [pelias/pelias#777]
Hey team! 

Is there any way currently to include post code information with locality results?

Checking the WOF heirachy, it unfortunately looks like postalcodes are regarded as lower than localities, and they have no children.
https://github.com/whosonfirst/whosonfirst-placetypes

However in Australia there is a very strict relationship where a postcode is a parent for localities, and localities will only ever have a single postcode. So it goes localadmin -> postalcode -> locality in the heirachy.

This difference has two effects:
a) Results that return a locality don't include any postalcode information
b) I have noticed some searches that include a postcode go a little funny (sorry no examples right now) and it's most reliable if I perform full address searches excluding the postcode.

**Comment by NickStallman** [comment]
Ah yes I did forget about CBDs which get a little weird for actual postal reasons. The rule generally is true beyond the CBDs.

Anyway it could be kept simple as the postcode is supplied with OpenAddresses (and G-NAF) for all of Australia. Or would PIP still be needed to get the internal WOF id? It should line up perfectly with whosonfirst-data-postalcode-au-latest.csv so it wouldn't need to muck around with polygons in theory.

**Comment by missinglink** [comment]
Another example I can think of outside the Melbourne CBD is `3043 Tullamarine` and `3045 Melbourne Airport`.

We haven't yet found a reliable way of adding missing postcode attributes to point-based records.

OA already has them in most cases for Australia, so I guess you're trying to do this for either OSM or some custom data?

**Comment by NickStallman** [comment]
Oh doh, yes sorry I completely forgot about that, postcodes are working perfectly for addresses but not suburb data sourced from WOF (i.e. for a "Search for a suburb" feature).
I clearly wasn't thinking when I wrote the last message - OpenAddresses/G-NAF won't help with that at all!

Would Pelias handle it if the WOF data included the postcode in the heirachy for the suburb by any chance? If so I can write a script to set that correctly when a suburb polygon fits completely inside a postcode polygon and contribute it back to WOF.

**Comment by NickStallman** [comment]
Using the official ABS shapefiles for postcodes and suburbs, I have dumped out a list of every single suburb that fits entirely within a postcode.
This was generated by a slightly crude script that imported

**Comment by NickStallman** [comment]
Ok so I've run a whole pile of comparisons and I have the following results.

Suburbs which exist entirely inside a postcode: 7545
Postcodes which fit entirely inside a suburb: 29

Comparisons were done using the ABS 2016 Suburb and Postcode datasets. The CSV files include their IDs.
All comparisons were done purely by polygon, and spot checks have confirmed the results are accurate.

I think this presents a strong case for including the postcode in the hierarchy for a locality in these cases where the relationship is strong.
I assume this would be done in WOF, and Pelias would just add the postcode if it exists in the hierachy? This would allow it to be used in Australia but not elsewhere where the relationship doesn't exist.

The postcodes which exist inside a suburb are what you'd expect, mostly Universities.

I've also included a list of suburbs which exist multiple times in a single state. There are actually a lot more than I was expecting.
In these cases the postcode is used to disambiguate which suburb the result is referring to.


[Australia Postcode⁄Suburb Comparison.zip](https://github.com/pelias/pelias/files/2862745/Australia.Postcode.Suburb.Comparison.zip)

**Comment by orangejulius** [comment]
Awesome analysis @NickStallman. I agree that special handling would have to be done somewhere, WOF or Pelias, for postal codes, since the relationship isn't the same in many other countries.

@stepps00 what would you think of adding postalcodes to Australian localities as described above?

**Comment by stepps00** [comment]
Thanks for the info, @NickStallman 

Who's On First has a handy list of [placetypes](https://github.com/whosonfirst/whosonfirst-placetypes) that each have certain "parent"/"parented by" rules which affect the way hierarchies are constructed. While some of the postalcodes may be within or encompass a locality (suburb), the locality will be in the postalcode hierarchy, not vice versa.

Looking at the postalcode for [3460](https://spelunker.whosonfirst.org/id/504779611/), it shares the same geometry as [Daylesford](https://spelunker.whosonfirst.org/id/101932829/). The postalcode correctly contains the Daylesford locality. In this case, it sounds like you're looking also include the postalcode of 3460 in the hierarchy of Daylesford..?

It could be possible to link these two records through another property (maybe `wof:coterminous` if geometries are the same), or some top-level property to indicate that.. but I don't think the changing locality hierarchies is the correct thing to do on the WOF end. 

Thoughts? I may need a bit more background about the use case.

**Comment by NickStallman** [comment]
@stepps00 yeah I have seen the placetypes graph. Unfortunately it isn't accurate for Australia. I had a interesting call with @missinglink where he let me know Australia was unusual, most countries use postal codes purely for postal reasons but in Australia they form part of the hierachy, between county and locality.

There will be cases where a suburb polygon equals a postcode polygon, but that's not what I'm referring to here. A postcode in Australia basically always has one or more suburbs inside it, and this issue is about storing the parent postcode for each suburb.

It's entirely normal to format a list of suburbs in Australia as "Suburb, Postcode" or "Suburb, Postcode, State" or "Suburb, State, Postcode"
Here are a couple of examples of this 
RealEstate.com.au
![image](https://user-images.githubusercontent.com/136553/52755486-3966ca00-3052-11e9-96e5-77241ad2047f.png)
Australia Post package delivery time estimation:
![image](https://user-images.githubusercontent.com/136553/52755560-82b71980-3052-11e9-87b3-55e79d459a5b.png)
Yellow Pages
![image](https://user-images.githubusercontent.com/136553/52755601-9b273400-3052-11e9-8525-4ee50049d90b.png)

The purpose of my checks was to verify that basically in every single instance Australian postcodes always contain one or more suburbs, and the postcode and suburb boundaries match.
E.g. if a postcode has 3 suburbs in it, the postcode's outline is essentially a Union of the suburb outlines.

My immediate goal is to allow Pelias to be able to produce similar results to the norm in Australia like what those screenshots show, which is currently not possible. It does make sense to have it in WOF however as it makes the data significantly more useful in Australia.

The postcode of a suburb is an important property to have over here. I'm not sure about the actual history of it however and why it differs from most other countries. I assume in other countries a similar format would be "locality, localadmin" or "locality, county" however localadmin and county are not used in that way in Australia. They aren't actually used very often for formatting addresses or when describing localities.

**Comment by NickStallman** [comment]
[missing_suburbs.zip](https://github.com/pelias/pelias/files/2866016/missing_suburbs.zip)
Here's a list of suburbs that don't fit exactly in a single postcode.

There are 7736 total.
6388 touch two or more postcode polygons.
1348 only touch a single postcode.

I'll see if I can find a separate authoriative source for the relationship beyond just polygon comparisons as well.

**Comment by rowanwins** [comment]
> but in Australia they form part of the hierachy, between county and locality.

I'm not sure this is entirely accurate... at least within the context of PSMA Admin Boundaries data. Attached is the [PSMA Administrative Boundaries Product Description](https://github.com/pelias/pelias/files/2867748/Administrative.Boundaries.Product.Description.pdf), this is a dataset that is compiled from various state government sources of data. 

On page 14-15 you get some helpful descriptions of the relationships inc 
"A locality has: 0 to many related Postcode records"

It's worth remembering that in Australia most admin boundaries are maintained by state govs while postcode regions are maintained independently by AusPost. As far as I'm aware there is mandate or mechanism to have them align. [Wikipedia](https://en.wikipedia.org/wiki/Postcodes_in_Australia) has some interesting info on anomalies.

Hope that helps

**Comment by NickStallman** [comment]
@rowanwins ahh thanks that's a bit more insight. So postcodes aren't actually a official government dataset (Federal/State/Local) rather it's Australia Post however the trend I've been trying to describe is that they must generally must follow suburb outlines.
That results them being commonly used in various ways such as what I highlighted.

This might make getting an official open licence postcode to suburb dataset a little tricky. :(
It looks like PSMA doesn't actually attempt to do it, and I believe the Australia Post datasets aren't free.

**Comment by missinglink** [comment]
Hi @NickStallman Happy New Year!

We've been looking at the relationships between postal codes and localities again recently and we now have these nice config files which might help with the discussion?

https://github.com/pelias/wof-admin-lookup/blob/be20ddf2857fbf07fd9450d54ef1b0070386a84b/src/data/AUS.tsv

related: https://github.com/pelias/wof-admin-lookup/pull/281

**Comment by NickStallman** [comment]
@missinglink awesome yeah I did notice some work there, although it looks like there's a bit of funny stuff floating around in the file?
- Some postcodes have the state appended, others don't. The first digit of the postcode relates to the state (0 = NT, 2 = NSW, 4 = QLD, etc...) so postcodes are unique nationally.
- There are some street names in the postcode field?
- Postcodes in Australia are very easy to validate, so stuff like "2???" doesn't get in. They should be 4 digits (leading 0 for NT) only, no more, no less and no other characters.

But apart from that, this should work extremely well for introducing the portcode to Pelias records.
Now that the lookup table is available, what's left to do to actually use it for populating query results?

**Comment by missinglink** [comment]
Yeah it's sourced from OSM so 🤷‍♂

**Comment by missinglink** [comment]
I believe we're trying to solve a different problem, but the lookup dataset might apply equally well to both.

What we're trying to solve is the issue of the PIP tool returning the locality name as defined by the local govt, which many people in the USA would prefer to see the 'Preferred Last Line' locality as defined by the USPS (who charge for the data).

I think if you'd like to use the same lookup tables to enrich data in Australia which is currently lacking a postal code (correct me if this isn't the topic of this issue), then there's two things to do:
- Test the mapping and see what level of quality we can get with it, how many false positives are introduced etc
- Discuss adding a new field called `postalcode_computed` (or some other name) which distinguishes algorithmically generated postalcodes from those which are sourced from authoritative means.

**Comment by missinglink** [comment]
If we can do this reliably I would suggest that the code lives within the PIP layer, so any importer can ask for the administrative areas which envelope a point and the service can reply with a mix of polygon intersections and table lookups, depending on the logic we define.

**Comment by NickStallman** [comment]
Yep that all sounds good to me! postcode_computed would be fine.

Where is the script that extracts it from OSM? I'll check it out and see how I can clean it up and also validate the quality vs a known good source.

**Comment by missinglink** [comment]
The data is generated with https://github.com/pelias/lastline

**Comment by sebfie** [comment]
Hello, I just created a pelias instance and I am surprised to not find a postalcode for a localadmin. On a autocomplete field, it's fair to be able to display the postal_code to help user choose beetween cities with the same name for example but different postal_codes. How to handle that?
